### PR TITLE
Compression/Decompression of G2 Points for BLS12_381

### PR DESCRIPTION
**Pull request by ColoCarletti**
# Compression/Decompression of G2 Points for BLS12_381

This PR fixes a bug in the function responsible for decompressing points on the BLS12_381 curve, where the 3rd bit was not being properly accounted for.

It also adds functionality for compressing points on this curve.

## Type of change

- [x] New feature
- [x] Bug fix

## Checklist
- [x] Unit tests added

### CODE REVIEW

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs [hunk 1]
```diff
@@ -106,6 +102,39 @@ impl Compress for BLS12381Curve {
...
+                FieldElement::new([x.value()[1].clone(), x.value()[0].clone()]);
+
+            let mut x_bytes = x_rev.to_bytes_be();
+
+            // Set first bit to to 1 indicate this is compressed element.
+            x_bytes[0] |= 1 << 7;
+
+            // Set the 3rd bit based on y value.
+            let y_neg = -y;
+            if y.value()[0].representative() > y_neg.value()[0].representative()
```
Comment (diegokingston) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs:129]: This condition should be y.value()[0].representative() > y_neg.value()[0].representative() || (y.value()[0].representative() == y_neg.value()[0].representative() && y.value()[1].representative() > y_neg.value()[1].representative()).
Perhaps a more efficient strategy is
y_is_greater_or_equal = y.value()[0].representative() >= y_neg.value()[0].representative().
if y_is_greater_or_equal || (y_is_greater_or_equal && y.value()[1].representative() > y_neg.value()[1].representative()

**Review by diegokingston — APPROVED** [review]

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs [hunk 2]
```diff
@@ -106,7 +102,48 @@ impl Compress for BLS12381Curve {
...
     }
 
-    #[allow(unused)]
+    #[cfg(feature = "alloc")]
+    fn compress_g2_point(point: &Self::G2Point) -> alloc::vec::Vec<u8> {
+        if *point == G2Point::neutral_element() {
+            // point is at infinity
+            let mut x_bytes = alloc::vec![0_u8; 96];
+            x_bytes[0] |= 1 << 7;
+            x_bytes[0] |= 1 << 6;
```
Comment (Oppen) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs:111]:
```suggestion
            x_bytes[0] |= 0b11 << 6;
```

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs [hunk 3]
```diff
@@ -106,7 +102,48 @@ impl Compress for BLS12381Curve {
             .ok_or(ByteConversionError::PointNotInSubgroup)
     }
 
-    #[allow(unused)]
+    #[cfg(feature = "alloc")]
+    fn compress_g2_point(point: &Self::G2Point) -> alloc::vec::Vec<u8> {
```
Comment (Oppen) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs:106]: Do we really need a `Vec` here or is it constant size?

Comment (diegokingston) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_381/compression.rs:106]: Traits get in the way, but we can try to solve this if we expose more things

**Review by Oppen — APPROVED** [review]

**Pull request merged**